package/MDAnalysis/coordinates/base.py

```diff
@@ -381,6 +385,15 @@ def from_coordinates(cls,
 
         return ts
 
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state.pop('_reader', None)
```

Comment:
I think it's fine to drop `_reader` after pickling. It seems that it is only being used to retrieve `dt` for some readers. Given that we have saved `dt` during pickling, it should never need to be used again.
```python
    def dt(self):
   2         """The time difference in ps between timesteps
   1             
851          Note
   1         ----
   2         This defaults to 1.0 ps in the absence of time data
   3 
   4 
   5         .. versionadded:: 0.11.0
   6         """
   7         try:
   8             return self.data['dt']
   9         except KeyError:
  10             pass
  11         try:
  12             dt = self.data['dt'] = self._reader()._get_dt()
  13             return dt
  14         except AttributeError:
  15             pass
  16         warnings.warn("Reader has no dt information, set to 1.0 ps")
  17         return 1.0
```

Reply:
So this only applies if we try to find `dt` (and trigger setting `self.data['dt']`) before pickling, otherwise we loose the file-based dt - e.g.
```python
    u = mda.Universe('testsuite/MDAnalysisTests/data/Amber/bala.ncdf')
    u_pickle_pre = pickle.loads(pickle.dumps(u))  # pickled before we check `dt`
    initial_dt = u.trajectory.dt
    u_pickle_post = pickle.loads(pickle.dumps(u))  # pickled after we check `dt`

    print('Initial dt: ', initial_dt)
    print('Pickle-post dt: ', u_pickle_post.trajectory.dt)
    print('Pickle-pre dt: ', u_pickle_pre.trajectory.dt)
```
returns `Initial dt: 0.0,    Pickle-post dt: 0.0,     Pickle-pre dt: 1.0` and warns that the default dt is being used for the pre case.

This could be avoided by calling dt so it's set before pickling? I guess technically what we get back then wouldn't be what we put in in terms of `data['dt']` now being set; I wouldn't think this is a huge deal, but looks like the copy methods in `Timestep` make the effort to pass `_reader` on rather than just make sure `dt` is set, and I'm not sure in general what the motivation is behind keeping the `_reader` weakref around to set dt only upon checking for it's value, rather than just doing this automatically during setup - maybe there is a good reason to keep it around...

Reply:
This is an interesting API. As @yuxuanzhuang mentions the reference to the reader is only used to calculate the time difference between two steps. We do not use it otherwise. The dt property is calculated by the [Timestep](https://github.com/MDAnalysis/mdanalysis/blob/8314f7d33b08bd467f1974fdc886f60b02f226c3/package/MDAnalysis/coordinates/base.py#L835) and exposed to the user through the [trajectory](https://github.com/MDAnalysis/mdanalysis/blob/8314f7d33b08bd467f1974fdc886f60b02f226c3/package/MDAnalysis/coordinates/base.py#L1460). @fiona-naughton is using `u.trajectory.dt` to get the the time difference.  

Do we really want to have dt be stored in the timestep? It would make more sense to me to only have access to dt from the trajectory. The timestep would only know about the current time, not the difference to another timestep. This would make logical sense to me as a time difference needs at least two timesteps to be defined. 

I would remove `dt` from the Timestep object and have the trajectory handle time difference calculation. We basically do this right now anyway, thats why we need the weakref to the trajectory (called reader in the code). This would make the need for this weak reference obsolete. 

I do not think this is a huge break of our public API. I never used a timestep object. I rather usually access the trajectory directly or an atomgroup.   Who actually accesses the timestep during iteration?

Other solution to the problem with lazy loading the time difference is to check for it when we call `__getstate__` before copying the dict

```python
def __getstate__(self):
    self.dt
    dict = self.__dict__.copy()
    dict.pop('_reader', None)
    return dict
```
The lazy loading logic is also the reason we keep the _reader reference around during a copy.

Reply:
> Who actually accesses the timestep during iteration?

`ts.frame` and `ts.time` are commonly used, I'd say. There are the various things in `ts.data`. Not sure about auxiliary stuff.

Dropping `ts.dt` makes sense, though.

Reply:
P.S.: There was some discussion/explanation in #2269 why we are _not_ really using `ts` a lot.

Reply:
@yuxuanzhuang can you open an issue for dropping `ts.dt`, please? It's an API change so it deserves its own issue and discussion. You can link it to this comment thread. Thanks.